Add support for pre of arrays to Cpp runtime

rfranke opened the pull request:
This appears needed as many examples of MSL are broken since the introduction of the new frontend.

DiscreteEvents/SimVars:
 - add pre of arrays

Array:
 - add new WrapArray for array return value from pre

FactoryConfig:
  - remove interface includes to avoid additional include of Array.h

CodegenCppCommon:
 - move $PRE qualifier for lhs from contextCref to cref1
 - make BaseArray function call arguments const

sjoelund commented:
It would probably be good to add a testcase for this so the same issue doesn't crop up in the future.

rfranke commented:
I added a test and fixed remaining issues that showed up with existing tests. Would be good to see the impact on the whole library coverage. Can you merge this PR?
